### PR TITLE
Homework 5 Submission

### PR DESCRIPTION
**Pull request by AleemRahil**

### CODE REVIEW

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Great job on your submission! You've done a good job of structuring your SQL query and your use of comments to explain your thought process is commendable. However, there are a few areas where your solution could be improved.

1) **Covering all aspects of the homework prompt:** Your solution does not fully cover all aspects of the homework prompt. The prompt asks for a state change-tracking field that takes on five different values based on the player's status. While your solution does account for these states, it does not correctly implement the logic for determining each state. For example, the 'Returned from Retirement' state should be assigned when a player was not active in the previous season but is active in the current season. However, your solution assigns this state when the difference between the current season and the last active season is greater than 1, which is not the same thing.

2) **Logical flow of the solution:** Your solution logically follows the given homework prompt. You've correctly used a FULL OUTER JOIN to combine the 'yesterday' and 'today' tables, and your use of the COALESCE function to handle NULL values is appropriate. However, the logic in your CASE statement for determining the 'yearly_active_state' could be improved. Consider revising this part of your solution to more accurately reflect the conditions outlined in the homework prompt.

3) **Readability of the code:** Your code is generally well-structured and easy to read. Your use of comments to explain your thought process is excellent. However, the variable names 'yesterday' and 'today' could be more descriptive. Consider renaming these to something like 'previous_season' and 'current_season' to more accurately reflect their purpose.

4) **Robustness and error handling:** Your solution does not handle all possible edge cases. For example, it does not account for the possibility that a player could be new to the league in the current season. In this case, the 'first_active_season' should be the current season, but your solution would assign a NULL value. Consider adding a condition to handle this case.

5) **Modularity and reusability of the code:** Your code is modular and could be reused with minor modifications. However, the logic for determining the 'yearly_active_state' is specific to the conditions outlined in the homework prompt and may not be applicable in other contexts. Consider revising this part of your solution to make it more generalizable.

Overall, you've done a good job on this assignment, but there are a few areas where your solution could be improved. Keep up the good work!

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

Overall, your solution is well-structured and logically sound. It correctly answers the question posed in the homework prompt, and your use of a Common Table Expression (CTE) and window function to rank players by their total points is a clever approach. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution does answer the question, but it doesn't exactly follow the prompt's instructions. The prompt asks you to build your query on top of the results of the `GROUPING SETS` aggregations, but your solution doesn't seem to use `GROUPING SETS` at all. While your solution does work, it would be better to follow the prompt's instructions more closely. 

2) **Logical coherence:** Your solution is logically sound, but it's a bit more complex than necessary. The example solution simply orders the results by `total_player_points` and takes the top result, which is a simpler and more straightforward approach. 

3) **Readability:** Your code is generally well-formatted and easy to read. Your use of a CTE and meaningful variable names makes your code easier to understand. However, your comments could be more descriptive. For example, you could explain why you're using a CTE and what the `DENSE_RANK()` function does.

4) **Robustness and error handling:** Your solution seems to handle edge cases well, such as players with no team or null player names. However, it's not clear how your solution would handle ties for the most points. The example solution doesn't handle this case either, but it's something to consider.

5) **Modularity and reusability:** Your use of a CTE makes your code more modular and reusable. However, your solution is more complex than necessary for this particular problem, which could make it harder to adapt for other uses.

In summary, your solution is a good start, but could be improved by following the prompt's instructions more closely, simplifying your approach, improving your comments, and considering how to handle ties for the most points.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully used window functions to calculate the most games a single team has won in a given 90-game stretch. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution covers most aspects of the homework prompt, but it seems to be missing the deduplication step that is present in the example solution. This step is important to ensure that each game is only counted once for each team. You can achieve this by adding a `ROW_NUMBER()` function in your `team_games` CTE, similar to the `deduped` CTE in the example solution.

2) **Readability:** Your code is generally well-formatted and easy to read. However, you could improve the readability by adding comments to explain your thought process, especially for complex parts of the code like the `CASE` statement in the `team_games` CTE and the window function in the `streak` CTE.

3) **Robustness and error handling:** Your code does not seem to handle potential errors or edge cases. For example, it does not account for the possibility of games being cancelled or postponed, which could affect the calculation of the 90-game stretch. You could add error handling code to account for these situations.

4) **Modularity and reusability:** Your code is structured in a way that makes it easy to understand and reuse specific components in other contexts. The use of CTEs (`team_games` and `streak`) helps to break down the problem into smaller, more manageable parts. However, you could improve the modularity of your code by encapsulating the `CASE` statement in the `team_games` CTE into a separate function, which could be reused in other queries.

Here's how you could modify your code to address these points:

```sql
WITH
 team_games AS ( 
   SELECT
    a.game_id,
    team_city,
    game_date_est,
    ROW_NUMBER() OVER(PARTITION BY a.game_id, a.team_id, player_id ORDER BY game_date_est) AS row_num,
    MAX(CASE WHEN (a.team_id = b.home_team_id AND b.home_team_wins = 1) OR
              (a.team_id = b.visitor_team_id AND b.home_team_wins = 0) THEN 1
    ELSE 0 END)
   AS games_won
   FROM
    bootcamp.nba_game_details a join bootcamp.nba_games b on a.game_id = b.game_id
   GROUP BY
     a.game_id,
     team_city,
     game_date_est
  
), streak AS (
  SELECT
   game_id,
   team_city,
   game_date_est,
   -- Calculate rolling sum over a 90 game window
   SUM(games_won) OVER(PARTITION BY team_city ORDER BY game_number
   ROWS BETWEEN 89 PRECEDING AND CURRENT ROW) AS rolling_games_won
 FROM
   team_games
 WHERE row_num = 1
)
SELECT
    team_city,
    MAX(rolling_games_won) AS streak_count
  FROM
   streak
 GROUP BY
   1
 ORDER BY
   2 DESC
```

Keep up the good work! With a few improvements, your code will be even better.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_7.sql`

The student's solution appears to cover all aspects of the homework prompt. The student has written a query to find out how many games in a row LeBron James scored over 10 points. The logic of the student's solution is sound and follows the given homework prompt. The student has used window functions and CTEs (Common Table Expressions) to break down the problem into manageable parts, which is a good practice.

The readability of the student's code is good. The student has used meaningful variable names and has included comments explaining their thought process, which makes the code easier to understand. The formatting of the code is also good, with proper indentation and line breaks, which makes the code easier to read.

The student's solution seems to handle edge cases and potential errors well. The student has used the `LAG` function to calculate the difference between the current row and the previous row, which helps to identify breaks between streaks. The student has also used the `CASE` function to handle cases where LeBron James scored less than or equal to 10 points.

The modularity and reusability of the student's code is good. The student has used CTEs to break down the problem into smaller parts, which makes the code easier to understand and reuse. The student has also used window functions, which are a powerful tool for solving complex SQL problems.

However, there are a few areas where the student's solution could be improved:

1) The student could add more comments to explain the logic behind their code. For example, the student could explain why they are using the `LAG` function and how it helps to identify breaks between streaks.

2) The student could improve the readability of their code by adding more line breaks and whitespace. For example, the student could add a line break before each `FROM` clause to make the code easier to read.

3) The student could improve the robustness of their code by adding error handling. For example, the student could add a `TRY...CATCH` block to handle potential errors.

4) The student could improve the modularity of their code by creating a function or stored procedure that encapsulates their logic. This would make the code easier to reuse in other contexts.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Your solution is well-structured and logically follows the given homework prompt. You have successfully used window functions to calculate the most games a single team has won in a given 90-game stretch. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution does not fully cover the homework prompt. The prompt asks for the most games a single team has won in a given 90-game stretch, but your solution calculates the most games won in a 90-day stretch. This is a significant difference, as there could be more or fewer than 90 games in a 90-day period. To fix this, you should change your window function to use `ROWS BETWEEN 89 PRECEDING AND CURRENT ROW` instead of `RANGE BETWEEN INTERVAL '89' DAY PRECEDING AND CURRENT ROW`.

2) **Readability:** Your code is generally well-formatted and easy to read. However, the variable names could be more descriptive. For example, `n_wins` could be renamed to `wins_in_90_games` to more accurately reflect its purpose. Also, consider adding more comments to explain your thought process and the purpose of each part of your code.

3) **Robustness and error handling:** Your code does not handle potential errors or edge cases. For example, it does not account for the possibility of games being cancelled or postponed, which could affect the number of games in a 90-day period. Consider adding error handling or checks to ensure the data is as expected.

4) **Modularity and reusability:** Your code is well-structured and modular, with each CTE building on the previous one. This makes it easy to understand and reuse specific components in other contexts. However, consider encapsulating parts of your code into functions or stored procedures for even greater reusability.

Overall, your solution is a good start, but with a few improvements, it could be even better. Keep up the good work!

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully used window functions to calculate the most games a single team has won in a given 90-game stretch. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution covers most aspects of the homework prompt, but it seems to be missing the part where you need to join the `teams` table to get the team name. Instead, you are returning the `team_id`. To improve this, you could join the `teams` table in your final query to get the `team_name` for the `team_id` with the most wins.

2) **Readability:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments to explain your thought process, which is great. However, the comment on the `deduped_games` CTE is a bit misleading. It seems to suggest that the `nba_games` table has duplicates, but you are not actually removing duplicates in this CTE. You are just selecting a subset of columns from the `nba_games` table. To improve readability, you could update this comment to more accurately reflect what the `deduped_games` CTE is doing.

3) **Robustness and error handling:** Your solution does not seem to handle potential errors or edge cases. For example, what happens if there are less than 90 games for a team? Or what if the `home_team_wins` column is null? To improve robustness, you could add some error handling or checks for these edge cases.

4) **Modularity and reusability:** Your solution is modular and the use of CTEs makes it easy to understand and reuse specific components. However, the `wins_over_90_days` CTE could be made more reusable by removing the `- interval '90' day` from the `window_start` calculation. This would allow you to easily change the window size without having to update this calculation.

Here's how you could update your final query to include the `team_name`:

```sql
select
    t.nickname as team_name,
    max_by(w.window_start, w.n_wins) as window_start,
    max_by(w.window_end, w.n_wins) as window_end,
    max(w.n_wins) max_wins_over_90_days
from
    wins_over_90_days w
join
    bootcamp.teams t on w.team_id = t.team_id
```

Keep up the good work! With a few minor adjustments, your solution will be even better.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully used window functions to calculate the most games a single team has won in a 90-game stretch. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution does not fully cover the homework prompt. The prompt asks for the most games a single team has won in a given 90-game stretch, but your solution calculates the most games won in a 90-day stretch. This is a significant difference, as there may not be a game every day. You should adjust your window function to count the number of games, not the number of days. 

2) **Readability:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments to explain your thought process, which is excellent. However, the comment on the `deduped_games` CTE is a bit misleading. It suggests that the `nba_games` table has duplicates, but it's not clear what kind of duplicates you're referring to. If you mean games that are listed twice (once for each team), it would be helpful to clarify this in your comment.

3) **Robustness and error handling:** Your code does not include any explicit error handling. While SQL queries do not typically include error handling in the same way as other programming languages, it's still important to consider potential issues that could arise. For example, what happens if the `nba_games` or `nba_game_details` tables are missing data for certain games or teams?

4) **Modularity and reusability:** Your code is well-structured and modular, with each CTE building on the previous one. This makes it easy to understand the overall flow of your solution. However, the `teams` CTE is only used to join the team nickname to the final result. It might be more efficient to include this join in the `wins_over_90_days` CTE, rather than creating a separate CTE for it.

Here's a suggestion for how you could adjust your window function to count games instead of days:

```sql
wins_over_90_games as (
    select
        team_id,
        game_date_est - interval '90' game as window_start,
        game_date_est as window_end,
        sum(is_win) over (
            partition by
                team_id
            order by
                game_date_est ROWS BETWEEN 89 PRECEDING AND CURRENT ROW
        ) as n_wins
    from
        games_both_perspectives
)
```

Keep up the good work, and with a few adjustments, your solution will be even better!

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully used window functions to calculate the most games a single team has won in a given 90-game stretch. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution covers most aspects of the homework prompt, but it seems to have misunderstood the requirement for a "90-game stretch". The prompt asks for the most games won in any 90-game stretch, not over a 90-day period. This means you should be looking at a rolling window of 90 games, not 90 days. You can achieve this by using `ROWS BETWEEN 89 PRECEDING AND CURRENT ROW` instead of `game_date_est - interval '90' day`.

2) **Readability:** Your code is generally well-formatted and easy to read. You've used meaningful variable names and included comments to explain your thought process, which is great. However, the comment on the `deduped_games` CTE is a bit misleading. It suggests that the `nba_games` table has duplicates, but your code doesn't actually remove any duplicates. It would be helpful to clarify this comment.

3) **Robustness and error handling:** Your code doesn't include any explicit error handling or checks for invalid inputs. While this might not be necessary for this specific assignment, it's a good habit to get into, especially when working with real-world data that can often be messy or inconsistent.

4) **Modularity and reusability:** Your use of CTEs (Common Table Expressions) makes your code modular and easier to understand. Each CTE is like a separate step in your overall process, which makes it easier to follow your logic and debug any issues. However, the `teams` CTE seems unnecessary, as you could join directly on the `nba_game_details` table in your final query.

Here's how you might revise your code to address these points:

```sql
with
    -- include games from both home team and away team's perspective 
    games_both_perspectives as (
        -- home team games
        select
            game_date_est,
            home_team_id as team_id,
            home_team_wins as is_win
        from
            bootcamp.nba_games
        union
        -- away team games
        select
            game_date_est,
            visitor_team_id as team_id,
            -- if was a win or not from away team's perspective
            case
                when home_team_wins = 1 then 0
                when home_team_wins = 0 then 1
            end as is_win
        from
            bootcamp.nba_games
    ),
    wins_over_90_games as (
    select
        team_id,
        game_date_est,
        sum(is_win) over (
            partition by
                team_id
            order by
                game_date_est ROWS BETWEEN 89 PRECEDING AND CURRENT ROW
        ) as n_wins
    from
        games_both_perspectives
    )
    
select
    t.team_abbreviation as team_name,
    max_by(w.game_date_est, w.n_wins) as window_end,
    max(w.n_wins) max_wins_over_90_games
from
    wins_over_90_games w
join
    bootcamp.nba_game_details t on w.team_id = t.team_id
GROUP BY
    t.team_abbreviation
```

Keep up the good work, and remember to carefully read the assignment prompt to ensure you're meeting all the requirements.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

Overall, your solution is well-structured and logically sound. You've successfully written a query that should return the team with the most wins. However, there are a few areas where your solution could be improved to better align with the assignment prompt and to enhance readability and robustness. Here are my suggestions:

1) **Alignment with the prompt:** The homework prompt asks for a query named `query_5`, but you haven't named your query. It's important to follow the instructions in the prompt exactly. You can name your query by adding a comment at the top, like this:

```sql
-- query_5: Find the team that has won the most games
```

2) **Readability:** Your code is generally well-formatted and easy to read. However, you could improve readability by adding more comments to explain your thought process. For example, you could explain why you're using a `WITH` clause and what the `DENSE_RANK()` function does.

3) **Robustness and error handling:** Your query includes a `WHERE` clause to exclude teams labeled 'N/A'. This is a good practice for handling invalid inputs. However, it's not clear from the prompt whether 'N/A' is the only invalid input. You might want to add error handling for other potential invalid inputs, such as NULL values.

4) **Modularity and reusability:** Your use of a `WITH` clause makes your query more modular and potentially reusable. However, the `WHERE n_r = 1` clause at the end of your query is a bit confusing. It seems like you're trying to get the top-ranked team, but it's not clear why you're using `n_r` instead of just ordering by `total_games_won` and using `LIMIT 1`, as in the example solution. If there's a specific reason for this, please explain it in a comment.

Here's how you might revise your query based on these suggestions:

```sql
-- query_5: Find the team that has won the most games
-- We're using a WITH clause to create a temporary table that ranks teams by total games won.
-- The DENSE_RANK() function assigns a rank to each team, with the highest total_games_won getting the lowest rank (1).
-- We then select the top-ranked team.
WITH
    ranked AS (
        SELECT
            team,
            total_games_won,
            DENSE_RANK() OVER(ORDER BY total_games_won DESC) AS n_r
        FROM
            aleemrahil84520.game_details_dashboard
        WHERE
            team <> 'N/A' AND team IS NOT NULL  -- Exclude invalid inputs
)
SELECT
  team,
  total_games_won,
  n_r
FROM
  ranked
WHERE
    n_r = 1  -- Select the top-ranked team
```

Keep up the good work! With a few minor adjustments, your solution will be even better.

### Feedback for `query_6.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used the `GROUPING SETS` clause to aggregate data along the required dimensions. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution does not fully cover the homework prompt. The prompt asks for aggregation along the dimensions of player and team, player and season, and team. However, in your `GROUPING SETS` clause, you have used `team_city` instead of `team_name` or `team_abbreviation`. Please ensure that you are using the correct column for team identification.

2) **Readability:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments to explain your thought process, which is good practice. However, you could improve the readability of your code by adding more detailed comments, especially for complex parts of your code like the `CASE` statement in the `SUM` function.

3) **Robustness and error handling:** Your code does not include any error handling or checks for invalid inputs. It would be good to add some checks to ensure that the input data is valid and to handle any potential errors that might occur. For example, you could add checks to ensure that the `player_name`, `team_city`, and `season` columns are not null before performing the aggregation.

4) **Modularity and reusability:** Your code is structured in a way that makes it easy to understand and reuse specific components in other contexts. The use of a `WITH` clause to create a temporary `combined` table is a good practice that enhances the modularity of your code. However, you could improve the reusability of your code by defining the aggregation logic in a separate function or stored procedure, which could then be reused in other queries.

In conclusion, your solution is on the right track, but there are a few areas where it could be improved. Keep up the good work and continue to refine your SQL skills!

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

File: submission/query_1.sql
Comment (HjersmanJ) [inline submission/query_1.sql:1]: Good job on code and creating header comments that describe approach and scope.

File: submission/query_2.sql
Comment (HjersmanJ) [inline submission/query_2.sql:1]: You did not alias the GROUPING SETS here. That is why they are not referenced in the following queries. This somewhat defeats the purpose of having them.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used the `GROUPING SETS` clause to aggregate data along the required dimensions. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution does not fully cover the homework prompt. The prompt asks for aggregation along the dimensions of player and team, player and season, and team. However, in your `GROUPING SETS` clause, you have used `team_city` instead of `team_name` or `team_abbreviation`. Please ensure that you are using the correct column for team information.

2) **Readability:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments to explain your thought process, which is good practice. However, you could improve the readability of your code by adding more comments to explain complex parts of your code, such as the `CASE` statement in the `SUM` function.

3) **Robustness and error handling:** Your code does not include any error handling or checks for invalid inputs. It would be good to add some checks to ensure that the data being processed is valid. For example, you could add checks to ensure that `player_name`, `team_city`, and `season` are not null before performing the aggregation.

4) **Modularity and reusability:** Your code is modular and the use of a CTE (Common Table Expression) makes it easy to understand and reuse specific components in other contexts. However, you could improve the reusability of your code by defining the aggregation logic in a separate function or stored procedure, which could then be reused in other queries.

5) **Syntax:** Please ensure that you are using the correct syntax for the `GROUPING` function. In the example solution, the `GROUPING` function is used with multiple arguments and a comparison operator to create a new column `aggregation_level`. This is not present in your solution. 

Here is a suggested modification to your code:

```sql
CREATE OR REPLACE TABLE aleemrahil84520.game_details_dashboard AS
WITH
  combined AS (
    SELECT 
     player_name AS player_name,
     COALESCE(team_abbreviation, 'N/A') AS team,
     season,
     SUM(a.pts) AS total_player_points,
     SUM(CASE
            WHEN (a.team_id = b.home_team_id AND home_team_wins = 1) OR
             (a.team_id = b.visitor_team_id AND home_team_wins = 0) THEN 1
             ELSE 0 END
     ) AS total_games_won
    FROM 
        bootcamp.nba_game_details_dedup AS a 
    JOIN
        bootcamp.nba_games AS b 
    ON 
        a.game_id = b.game_id
    WHERE player_name IS NOT NULL AND team_abbreviation IS NOT NULL AND season IS NOT NULL
    GROUP BY
     GROUPING SETS(
       (player_name, team_abbreviation),
       (player_name, season),
       (team_abbreviation)
     )
)
SELECT
    *
FROM
    combined
```

Keep up the good work and continue practicing your SQL skills!

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

File: submission/query_3.sql [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    ranked AS (
+        SELECT
+            player_name,
+            team,
+            total_player_points,
+            DENSE_RANK() OVER(ORDER BY total_player_points DESC) AS n_r
+        FROM
+            aleemrahil84520.game_details_dashboard
+        WHERE
+            team <> 'N/A'
```
Comment (HjersmanJ) [inline submission/query_3.sql:13]: While this grabs the data, it makes it far harder to read for analytic consumers. In essence, a grouping set would be calling each set as if it were a table itself and finding the aggregation level depending on a pre-defined alias. So while correct, it doesn't use grouping sets as intended.

File: submission/query_6.sql [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                when home_team_wins = 1 then 0
+                when home_team_wins = 0 then 1
+            end as is_win
+        from
+            deduped_games
+    ),
+    wins_over_90_days as (
+    select
+        team_id,
+        game_date_est - interval '90' day as window_start,
```
Comment (HjersmanJ) [inline submission/query_6.sql:42]: Since you already are checking the correct range on line 48, the affine interval shift is not necessary.

File: submission/query_6.sql
Comment (HjersmanJ) [inline submission/query_6.sql:1]: You need to order it by the games won descending and pick the top 1.

File: submission/query_7.sql [hunk 3]
```diff
@@ -0,0 +1,48 @@
+-- Find the highest streak of games where
+-- LeBron James has score over t0 points
```
Comment (HjersmanJ) [inline submission/query_7.sql:2]: ...scored over *10* points. Just minor error in comment.

File: submission/query_7.sql
Comment (HjersmanJ) [inline submission/query_7.sql:1]: This query produces 136 for me . Correct answer is 62. I suspect you have data that is not deduped that is ruining the output.

**Review by HjersmanJ — APPROVED** [review]
A few errors in the code that produces erroneous outputs. You also are not aliasing the grouping sets as intended, which makes an analyst's job harder to understand which one to use. Many people do not want to check if a certain row is 0.

Since you attempted all of the queries, you have shown sufficient knowledge to pass you for this assignment. Good job!
